Python interface for updating translation and rotation rates of markers

HahsFilip opened the pull request:
## Proposed Changes
*Give a brief overview of your contribution here in a few sentences.*
 My changes allow the changing of moving wall velocities when using the py_wrapper. 

## Related Work
*Resolve any issues (bug fix or feature request), note any related PRs, or mention interactions with the work of others, if any.*
There has been some previous work, where functionality for changing mesh rotation rates was added.


## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [X ] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [x] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

HahsFilip commented:
@pcarruscag This is my first PR, so I hope that at least a few things will be ok :)

HahsFilip commented on TestCases/py_wrapper/rotating_cylinder/run.py:
```diff
@@ -0,0 +1,30 @@
+import pysu2ad
+from mpi4py import MPI
+import numpy as np
+import shutil
```
Resolves

HahsFilip commented on TestCases/py_wrapper/rotating_cylinder/run.py:
```diff
@@ -0,0 +1,30 @@
+import pysu2			            # imports the SU2 wrapped module
+from math import *
+import os
+import pysu2ad
```
Resolved

HahsFilip commented on TestCases/py_wrapper/rotating_cylinder/run.py:
```diff
@@ -0,0 +1,30 @@
+from optparse import OptionParser	# use a parser for configuration
+import pysu2			            # imports the SU2 wrapped module
+from math import *
+import os
```
Resolved

HahsFilip commented on TestCases/py_wrapper/rotating_cylinder/run.py:
```diff
@@ -0,0 +1,30 @@
+import sys
+from optparse import OptionParser	# use a parser for configuration
```
Resolved

HahsFilip commented on TestCases/py_wrapper/rotating_cylinder/run.py:
```diff
@@ -0,0 +1,30 @@
+import sys
```
Resolved

pcarruscag commented on Common/include/CConfig.hpp:
```diff
@@ -5962,6 +5971,16 @@ class CConfig {
+   * \param[in] iDim - spatial component
+   * \param[in] val - Rotational velocity
+   */
+  void SetMarkerRotation_Rate(unsigned short iMarkerMoving, unsigned short iDim, su2double val) {
```
```suggestion
  void SetMarkerRotationRate(unsigned short iMarkerMoving, unsigned short iDim, su2double val) {
```

pcarruscag commented on SU2_CFD/src/python_wrapper_structure.cpp:
```diff
@@ -167,3 +167,19 @@ void CDriver::SetRotationRate(passivedouble rot_x, passivedouble rot_y, passived
 
+void CDriver::SetMarkerRotationRate(unsigned short iMarker, passivedouble rot_x, passivedouble rot_y, passivedouble rot_z) {
+  for (iZone = 0; iZone < nZone; iZone++) {
+    config_container[iZone]->SetMarkerRotation_Rate(iMarker, 0, rot_x);
+    config_container[iZone]->SetMarkerRotation_Rate(iMarker, 1, rot_y);
+    config_container[iZone]->SetMarkerRotation_Rate(iMarker, 2, rot_z);
+  }
+}
+
+void CDriver::SetMarkerTranslationRate(unsigned short iMarker, passivedouble vel_x, passivedouble vel_y, passivedouble vel_z) {
+  for (iZone = 0; iZone < nZone; iZone++) {
+    config_container[iZone]->SetMarkerTranslation_Rate(iMarker, 0, vel_x);
+    config_container[iZone]->SetMarkerTranslation_Rate(iMarker, 1, vel_y);
+    config_container[iZone]->SetMarkerTranslation_Rate(iMarker, 2, vel_z);
```
```suggestion
    config_container[iZone]->SetMarkerRotationRate(iMarker, 0, rot_x);
    config_container[iZone]->SetMarkerRotationRate(iMarker, 1, rot_y);
    config_container[iZone]->SetMarkerRotationRate(iMarker, 2, rot_z);
  }
}

void CDriver::SetMarkerTranslationRate(unsigned short iMarker, passivedouble vel_x, passivedouble vel_y, passivedouble vel_z) {
  for (iZone = 0; iZone < nZone; iZone++) {
    config_container[iZone]->SetMarkerTranslationRate(iMarker, 0, vel_x);
    config_container[iZone]->SetMarkerTranslationRate(iMarker, 1, vel_y);
    config_container[iZone]->SetMarkerTranslationRate(iMarker, 2, vel_z);
```

pcarruscag commented on TestCases/py_wrapper/rotating_cylinder/run.py:
```diff
@@ -0,0 +1,26 @@
+import pysu2			            # imports the SU2 wrapped module
```
Thanks :+1: to make a regression test from this you need to add it to parallel_regression.py

pcarruscag replied:
Add our license header to this file please (since it's also code).

kursatyurt commented on TestCases/py_wrapper/rotating_cylinder/run.py:
```diff
@@ -0,0 +1,26 @@
+import numpy as np
+
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+ad_com = MPI.COMM_WORLD
+rank_ad = ad_com.Get_rank() 
```
```suggestion
```

Unused lines

HahsFilip replied:
Removed

kursatyurt commented on TestCases/py_wrapper/rotating_cylinder/run.py:
```diff
@@ -0,0 +1,26 @@
+import pysu2			            # imports the SU2 wrapped module
+from math import *
```
```suggestion
```

This is not used as well.

HahsFilip commented:
I'm having some trouble getting the regression tests to pass - for some reason there is a result difference in the moving wall tests. I'm not sure how this can be possible, considering I am not changing any values.  The convergence seems to be slowed down considerably, but I'm having trouble identifying which function may be causing this.

pcarruscag commented:
Why do you say the convergence is slowed down? It looks like a very small change. Did you run the cases for more iterations?

HahsFilip commented:
Well, I've noticed the problem, because the unit tests don't pass - cavity and spinning cylinder especially. They exceed the number of iterations, and at the same time there is a result discrepancy.

HahsFilip commented:
I'm actually having the same problems running the code on the develop branch....

HahsFilip commented:
So you mean I should run the tests to convergence, and compare the results between branches? I wonder why compilation problems could be the culprit, as the tests fail both locally and in the github test. Shouldn't the github test always be compiled the same way?

pcarruscag commented:
You mentioned issues with the develop branch, those are likely due to compilation differences.
Yes run to convergence with both branches and compare please. I expect integrated quantities will be exactly the same.

pcarruscag commented on SU2_CFD/include/drivers/CDriver.hpp:
```diff
@@ -535,7 +535,24 @@ class CDriver : public CDriverBase {
    * \param[in] rot_z - Value of Angular velocity about z-axes.
    */
   void SetRotationRate(passivedouble rot_x, passivedouble rot_y, passivedouble rot_z);
+  /*!
```
```suggestion

  /*!
```

pcarruscag commented on SU2_CFD/src/python_wrapper_structure.cpp:
```diff
@@ -167,3 +167,19 @@ void CDriver::SetRotationRate(passivedouble rot_x, passivedouble rot_y, passived
 }
 
+void CDriver::SetMarkerRotationRate(unsigned short iMarker, passivedouble rot_x, passivedouble rot_y, passivedouble rot_z) {
+  for (iZone = 0; iZone < nZone; iZone++) {
+    config_container[iZone]->SetMarkerRotationRate(iMarker, 0, rot_x);
+    config_container[iZone]->SetMarkerRotationRate(iMarker, 1, rot_y);
+    config_container[iZone]->SetMarkerRotationRate(iMarker, 2, rot_z);
+  }
+}
+
+void CDriver::SetMarkerTranslationRate(unsigned short iMarker, passivedouble vel_x, passivedouble vel_y, passivedouble vel_z) {
+  for (iZone = 0; iZone < nZone; iZone++) {
+    config_container[iZone]->SetMarkerTranslationRate(iMarker, 0, vel_x);
+    config_container[iZone]->SetMarkerTranslationRate(iMarker, 1, vel_y);
+    config_container[iZone]->SetMarkerTranslationRate(iMarker, 2, vel_z);
+  }
```
```suggestion
  config_container[selected_zone]->SetMarkerRotationRate(iMarker, 0, rot_x);
  config_container[selected_zone]->SetMarkerRotationRate(iMarker, 1, rot_y);
  config_container[selected_zone]->SetMarkerRotationRate(iMarker, 2, rot_z);
}

void CDriver::SetMarkerTranslationRate(unsigned short iMarker, passivedouble vel_x, passivedouble vel_y, passivedouble vel_z) {
  config_container[selected_zone]->SetMarkerTranslationRate(iMarker, 0, vel_x);
  config_container[selected_zone]->SetMarkerTranslationRate(iMarker, 1, vel_y);
  config_container[selected_zone]->SetMarkerTranslationRate(iMarker, 2, vel_z);
```

pcarruscag reviewed (commented):
Hey @HahsFilip do you think you can wrap this up?
Take care of Kursat's comments and update the failing tests since it seem to be just numerical noise (or do you have reason to believe otherwise?)

pcarruscag commented:
You also need to add yourself to the AUTHORS.md file

HahsFilip commented:
@pcarruscag, sorry for the delay I had to get my life together a bit after the conference :). I think I have the correct changes, I have changed the tests and removed unused lines.

Filip

HahsFilip commented on TestCases/py_wrapper/rotating_cylinder/run.py:
```diff
@@ -0,0 +1,30 @@
+rank_ad = ad_com.Get_rank() 
+n_of_steps = 10
+rotation_vector = np.zeros(n_of_steps)+20
+rotation_vector[5:9] = -rotation_vector[5:9]
```
Done

pcarruscag reviewed (approved):
LGTM

pcarruscag commented on externals/codi:
You need to revert these changes to submodules, this may be why you see some failing cases

kursatyurt commented:
@pcarruscag , @HahsFilip 

Do the hybrid tests need an update? Do any of you have any idea about why they are failing?

pcarruscag reviewed (commented):
Because this branch has the wrong versions of various submodules

kursatyurt commented:
> Because this branch has the wrong versions of various submodules

I  see I assumed the last commit reverted all. As it is on a fork we cannot revert 👍🏼

kursatyurt commented:
@pcarruscag, is it possible to merge this PR to some temporary branch in the SU2 repo, fix the broken submodules there, and merge it to the main branch? It looks like @HahsFilip is busy. PR is almost complete but cannot be merged. I will do the required fixes.

pcarruscag commented:
Thanks @kursatyurt submodules are fixed here https://github.com/su2code/SU2/commit/8b89e857ff72782b0caccc3f63385c03bc67f6f7
If you can do the other fixes and hook up the regression test it would be great

kursatyurt commented:
Thanks @pcarruscag , it looks like standard regression tests are passed without any problem, and ARM tests are problematic, as always.
